E0107 needs to be updated to new format

From: src/test/compile-fail/E0107.rs

Error E0107 needs a span_label, updating it from:

```
error[E0107]: wrong number of lifetime parameters: expected 1, found 0
  --> src/test/compile-fail/E0107.rs:20:10
   |
20 |     foo: Foo, //~ ERROR E0107
   |          ^^^

error[E0107]: wrong number of lifetime parameters: expected 0, found 1
  --> src/test/compile-fail/E0107.rs:21:10
   |
21 |     bar: Bar<'a>, //~ ERROR E0107
   |          ^^^^^^^
```

To:

```
error[E0107]: wrong number of lifetime parameters: expected 1, found 0
  --> src/test/compile-fail/E0107.rs:20:10
   |
20 |     foo: Foo, //~ ERROR E0107
   |          ^^^ expected 1 lifetime parameter

error[E0107]: wrong number of lifetime parameters: expected 0, found 1
  --> src/test/compile-fail/E0107.rs:21:10
   |
21 |     bar: Bar<'a>, //~ ERROR E0107
   |          ^^^^^^^ unexpected lifetime parameter
```

This assumes we can see how many lifetime parameters we expect, and update the label accordingly.

--- a/src/librustc_typeck/astconv.rs
+++ b/src/librustc_typeck/astconv.rs
@@ -2266,9 +2266,25 @@ fn check_type_argument_count(tcx: TyCtxt, span: Span, supplied: usize,
 }
 
 fn report_lifetime_number_error(tcx: TyCtxt, span: Span, number: usize, expected: usize) {
-    span_err!(tcx.sess, span, E0107,
-              "wrong number of lifetime parameters: expected {}, found {}",
-              expected, number);
+    let label = if number < expected {
+        if expected == 1 {
+            format!("expected {} lifetime parameter", expected)
+        } else {
+            format!("expected {} lifetime parameters", expected)
+        }
+    } else {
+        let additional = number - expected;
+        if additional == 1 {
+            "unexpected lifetime parameter".to_string()
+        } else {
+            format!("{} unexpected lifetime parameters", additional)
+        }
+    };
+    struct_span_err!(tcx.sess, span, E0107,
+                     "wrong number of lifetime parameters: expected {}, found {}",
+                     expected, number)
+        .span_label(span, &label)
+        .emit();
 }
 
 // A helper struct for conveniently grouping a set of bounds which we pass to

--- a/src/test/compile-fail/E0107.rs
+++ b/src/test/compile-fail/E0107.rs
@@ -17,8 +17,12 @@ enum Bar {
 }
 
 struct Baz<'a> {
-    foo: Foo, //~ ERROR E0107
-    bar: Bar<'a>, //~ ERROR E0107
+    foo: Foo,
+    //~^ ERROR E0107
+    //~| expected 1 lifetime parameter
+    bar: Bar<'a>,
+    //~^ ERROR E0107
+    //~| unexpected lifetime parameter
 }
 
 fn main() {

--- a/src/test/compile-fail/E0107.rs
+++ b/src/test/compile-fail/E0107.rs
@@ -9,6 +9,7 @@
 // except according to those terms.
 
 struct Foo<'a>(&'a str);
+struct Buzz<'a, 'b>(&'a str, &'b str);
 
 enum Bar {
     A,
@@ -16,13 +17,19 @@ enum Bar {
     C,
 }
 
-struct Baz<'a> {
+struct Baz<'a, 'b, 'c> {
     foo: Foo,
     //~^ ERROR E0107
     //~| expected 1 lifetime parameter
+    buzz: Buzz<'a>,
+    //~^ ERROR E0107
+    //~| expected 2 lifetime parameters
     bar: Bar<'a>,
     //~^ ERROR E0107
     //~| unexpected lifetime parameter
+    foo2: Foo<'a, 'b, 'c>,
+    //~^ ERROR E0107
+    //~| 2 unexpected lifetime parameters
 }
 
 fn main() {

